### PR TITLE
Fixed issue #16853: Inconsistent filter behaviour when create token

### PR DESCRIPTION
**Pull request by Shnoulle**
Dev: remove flattenText in controller
Dev: usage of filter_var on attribute for all user (admin or not)

### CODE REVIEW

File: application/models/Token.php [hunk 1]
```diff
@@ -366,7 +383,11 @@ public function rules()
             array('emailstatus', 'default', 'value' => 'OK'),
         );
         foreach (decodeTokenAttributes($this->survey->attributedescriptions) as $key => $info) {
-                $aRules[] = array($key, 'LSYii_Validators', 'except'=>'FinalSubmit');
+            $aRules[] = array(
+                $key, 'filter',
+                'filter' => array(self::class, 'sanitizeAttribute'),
+                'except'=>'FinalSubmit'
```
Comment (Shnoulle) [inline application/models/Token.php:389]: I muts add `'except'=>'FinalSubmit'` for firstname or lastname too ? It's a sanitizer, not a checker.

File: application/models/Token.php [hunk 2]
```diff
@@ -263,6 +263,15 @@ public static function sanitizeToken($token)
...
     }
 
+    /**
+     * Sanitize string for any attribute
+     * @param string $attribute to sanitize
+     * @return string sanitized attribute
+     */
+    public static function sanitizeAttribute($attribute)
+    {
+        return filter_var($attribute, FILTER_SANITIZE_STRING);
```
Comment (Shnoulle) [inline application/models/Token.php:273]: With a checker (is valid) `return filter_var($attribute, FILTER_SANITIZE_STRING) == $attribute` maybe bettre : don't update, just send a error and disable save.
But : unsure admin GUI was ready, same with import …

**Comment by Shnoulle** [comment]
See https://bugs.limesurvey.org/view.php?id=16884 :)

File: application/controllers/admin/tokens.php [hunk 3]
```diff
@@ -541,17 +541,17 @@ public function editToken($iSurveyId)
 
             echo $from . ',' . $until;
             $aData = array(
-                'firstname' => flattenText(Yii::app()->request->getPost('firstname')),
```
Comment (olleharstedt) [inline application/controllers/admin/tokens.php:544]: Why is flattenText removed?

Comment (Shnoulle) [reply application/controllers/admin/tokens.php:544]: Because it was ONLY for admin … via GUI : not for

- register
- remote control

for example.

Question : did you read the mantis issue ?

Comment (Shnoulle) [reply application/controllers/admin/tokens.php:544]: With this one we flatten for all action (using model …) `return filter_var($attribute, FILTER_SANITIZE_STRING);` not **only** for admin … user via GUI

Comment (Shnoulle) [reply application/controllers/admin/tokens.php:544]: See related : https://bugs.limesurvey.org/view.php?id=16884 …

Comment (Shnoulle) [reply application/controllers/admin/tokens.php:544]: CPDB allow tag too

Comment (olleharstedt) [reply application/controllers/admin/tokens.php:544]: OK :)

Comment (Shnoulle) [reply application/controllers/admin/tokens.php:544]: Can resolve ?

File: application/models/Token.php (hunk 2 above)
Comment (Shnoulle) [reply application/models/Token.php:273]: Check : admin GUI not ready …

**Comment by olleharstedt** [comment]
Did anyone test this?

**Comment by Shnoulle** [comment]
Me …

**Comment by olleharstedt** [comment]
The dev should not test his/her own code. :)

**Comment by olleharstedt** [comment]
At least not only the dev. :d

**Comment by Shnoulle** [comment]
;)

Here : see the original issue one the [«way to reproduce»](https://bugs.limesurvey.org/view.php?id=16853). But need 3 seperate test (minimum 3)

But this fix fix https://bugs.limesurvey.org/view.php?id=16884 too …

**Comment by c-schmitz** [comment]
Denix, can you port this also to master and develop, please?

**Comment by Shnoulle** [comment]
Done on both

**Comment by olleharstedt** [comment]
Seems like this fix is incorrectly escaping `'` (apostrophe) in participant names.

**Comment by olleharstedt** [comment]
Possible fix here: https://github.com/LimeSurvey/LimeSurvey/commit/b4953e4a4c15db37fca875c3b21c59a6eaf31156

**Comment by Shnoulle** [comment]
Thank you
